crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -481,6 +578,60 @@ fn vm_move_call<E: fmt::Debug, S: StorageView<E>>(
     Ok(result)
 }
 
+fn upgrade_check_compatiblity_and_publish_package<E: fmt::Debug, S: StorageView<E>>(
+    context: &ExecutionContext<E, S>,
+    current_package: &MovePackage,
+    upgraded_package_modules: Vec<CompiledModule>,
+) -> Result<Object, ExecutionError> {
+    let mut upgraded_package_object = Object::new_package(
+        upgraded_package_modules,
+        current_package.version(),
+        context.tx_context.digest(),
+        context.protocol_config.max_move_package_size(),
+    )?;
+
+    let upgraded_pkg = upgraded_package_object
+        .data
+        .try_as_package_mut()
+        .ok_or_else(|| ExecutionError::from_kind(ExecutionErrorKind::InvariantViolation))?;
+
+    // Check compatibility
+
+    let check_struct_and_pub_function_linking = true;
+    let check_struct_layout = true;
+    let check_friend_linking = false;
+    let compatibility = Compatibility::new(
+        check_struct_and_pub_function_linking,
+        check_struct_layout,
+        check_friend_linking,
+    );
+
+    // The current package should be able to be normalized otherwise something has gone terribly wrong.
+    let current_normalized = current_package
+        .normalize()
+        .map_err(|_| ExecutionError::from_kind(ExecutionErrorKind::InvariantViolation))?;
+    let mut new_normalized = upgraded_pkg.normalize().map_err(|_| {
+        ExecutionError::from_kind(ExecutionErrorKind::PackageUpgradeError(
+            PackageUpgradeError::IncompatibleUpgrade,
+        ))
+    })?;
+
+    for (name, cur_module) in current_normalized {
+        let Some(new_module) = new_normalized.remove(&name) else {
+            return Err(ExecutionError::from_kind(ExecutionErrorKind::PackageUpgradeError(PackageUpgradeError::IncompatibleUpgrade)))
+        };
+
+        if compatibility.check(&cur_module, &new_module).is_err() {
```

Comment:
What is the error here? Does it give more info? Should we pass that along to the user?
Maybe at least we could give it as a string for dev-inspect cases

Reply:
`ExecutionError::new_with_source` is the function for doing so